iocore/utils/PropertyBlock.h

```diff
@@ -0,0 +1,254 @@
+
+#include "stdint.h"
+#include <cstddef>
+#include <atomic>
+#include <cstring>
+#include <vector>
+#include <type_traits>
+
+#include "ts/ink_assert.h"
+
+/// runtime defined static container
+/** The size of this structure is actually zero, so it will not change the size of your derived class.
+ * But new and delete are overriden to use allocate enough bytes of the derived type + properties added
+ * at run time using ioBufferAllocator.
+ *
+ * Also all bools are packed to save space using the *Bit methods.
+ *
+ * This is templated so static variables are instanced per Derived type. B/c we need to have different
+ * size property sets.
+ */
+template <class Derived_t> ///< static members instance pey type.
+struct PropertyBlock {
+public:
+  using self_t = PropertyBlock<Derived_t>;
+  /// callback format used to PropBlock init and destroy
+  using PropertyFunc_t = void(Derived_t *, void *);
+
+  /// number of bytes from 'this', or the bitset index
+  using Offset_t = uint32_t;
+
+  /// Add a data type to schema.
+  /**
+   * @param prop_count the number of instances of this data type to allocate.
+   * @param init an initialization function for property
+   * @param destroy a destruction function for property
+   * Note: if prop_count is 0, the init and destroy are still called per instance.
+   */
+  template <typename Property_t>
+  static Offset_t
+  PropBlockDeclare(size_t prop_count, PropertyFunc_t *init, PropertyFunc_t *destroy)
+  {
+    ink_assert(s_instance_count == 0); // it's too late, we already started allocating.
+    static_assert(std::is_same<Property_t, bool>::value == false,
+                  "Use PropBlockDeclareBit so we can pack bits, and we can't return a pointer to a bit.");
+
+    Offset_t offset = sizeof(Derived_t) + s_properties_total_size;
+    s_properties_total_size += prop_count * sizeof(Property_t);
+    if (init || destroy) {
+      for (Offset_t head = offset; prop_count > 0; head += sizeof(Property_t)) {
+        getSchema().push_back((Block){head, init, destroy});
+        --prop_count;
+      }
+    }
+    return offset;
+  }
+
+  template <class Property_t>
+  static Offset_t
+  PropBlockDeclare(size_t prop_count)
+  {
+    static auto default_init    = [](Derived_t *, void *ptr) { new (ptr) Property_t; };
+    static auto default_destroy = [](Derived_t *, void *ptr) { reinterpret_cast<Property_t *>(ptr)->~Property_t(); };
+    // use default constructors of class
+    return PropBlockDeclare<Property_t>(prop_count, default_init, default_destroy);
+  }
+
+  template <typename Property_t>
+  Property_t const &
+  propRead(Offset_t offset) const
+  {
+    ink_assert(Derived_t::hasReadAccess(static_cast<const Derived_t *>(this))); // the Derived_t probably needs a thread lock
+    ink_assert(propGetBit(initialized) == true); // the Derived_t needs to call propBlockInit() during construction.
+    return *static_cast<Property_t *>(propBlockGetPtr(offset));
+  }
+
+  template <typename Property_t>
```

Comment:
I don't see the point of distinguishing these. If the caller cares, he can simply declare a `const&` on his side.

Reply:
It only matters if we use a read-write lock. I considered making the read write operation implicit based on the constness, IE `const prop* <type>get(offset) const` and `prop* <type>get(offset)`. But I was uncomfortable making multi-threading functionality so transparent.